# RNAseq_deconvolution
1. The scripts provided here are served as a example for the deconvolution of VB-SE mixtures, but could be extended to other multi-body-fluid-mixtures by revise the name of the folders from VB or SE to any body fluids existed in mixtures.
2. sample-ex indicates the name of mixture sample.
3. The bed files used in the scripts should include the information of chromosome, start position, end position and gene symbol of relative specific genes, for example, for VB-SE mixtures, "VB_specific.bed" should including the above information of overexpressed genes in venous blood compared to semen. 
4. The vcf files "sample-ex.dpmaf.recode.vcf" and "sample-ex.EAS.dpmaf.recode.vcf" are obtained from GATK-SNP calling result of the mixture and variant calling data accessed from 1000 genomic Phase 3 datasets. The former includes the SNPs and their genotypes detected in mixtures filtered by DP and GQ, excluding the SNPs located in imprinted genes or with RNA editing events recorded in current databases, the latter includes the genotype informations of these SNPs of 504 East Asian peoples from the 1000 Genome projects. The SNPs in "sample-ex.dpmaf.recode.vcf" and "sample-ex.EAS.dpmaf.recode.vcf" should be completely the same otherwise the phenotype acquired from the script "mixture_phenotype.sh" could not be assigned to a phenotype frequency calculated by the script "phenotype_frequency.sh".
5. The python scripts "summarized_data.py" were used to automatically assign the mixture phenotype to a phenotype frequency, which could be implemented by the code (a example for the results of VB contributor):
   python summarized_data.py -m sample-ex.VBmix_pheno.txt -e sample-ex.VB_EAS_phenofreq.txt -o sample-ex.VB.xlsx
in this code, the "sample-ex.VBmix_pheno.txt" is the result of VB contributor accessed by "mixture_phenotype.sh", and the "sample-ex.VB_EAS_phenofreq.txt" is the result of VB contributor accessed by "phenotype_frequency.sh".
